azure-iot-device/tests/iothub/pipeline/test_pipeline_stages_iothub.py

```diff
@@ -378,6 +383,364 @@ def test_update_fails(
+        stage = cls_type(**init_kwargs)
+        stage.send_op_down = mocker.MagicMock()
+        stage.send_event_up = mocker.MagicMock()
+        return stage
+
+
+class EnsureDesiredPropertiesStageInstantiationTests(EnsureDesiredPropertiesStageTestConfig):
+    @pytest.mark.it("Initializes 'last_version_seen' None")
+    def test_last_version_seen(self, init_kwargs):
+        stage = pipeline_stages_iothub.EnsureDesiredPropertiesStage(**init_kwargs)
+        assert stage.last_version_seen is None
+
+    @pytest.mark.it("Initializes 'pending_get_request' None")
+    def test_pending_get_request(self, init_kwargs):
+        stage = pipeline_stages_iothub.EnsureDesiredPropertiesStage(**init_kwargs)
+        assert stage.pending_get_request is None
+
+
+pipeline_stage_test.add_base_pipeline_stage_tests(
+    test_module=this_module,
+    stage_class_under_test=pipeline_stages_iothub.EnsureDesiredPropertiesStage,
+    stage_test_config_class=EnsureDesiredPropertiesStageTestConfig,
+    extended_stage_instantiation_test_class=EnsureDesiredPropertiesStageInstantiationTests,
+)
+
+
+@pytest.mark.describe(
+    "EnsureDesiredPropertiesStage - .run_op() -- Called with EnableFeatureOperation"
+)
+class TestEnsureDesiredPropertiesStageRunOpWithEnableFeatureOperation(
+    StageRunOpTestBase, EnsureDesiredPropertiesStageTestConfig
+):
+    @pytest.fixture
+    def op(self, mocker):
+        return pipeline_ops_base.EnableFeatureOperation(
+            feature_name="fake_feature_name", callback=mocker.MagicMock()
+        )
+
+    @pytest.mark.it("Sets `last_version_seen` to -1 if `op.feature_name` is 'twin_patches'")
+    def test_sets_last_version_seen(self, mocker, stage, op):
+        op.feature_name = pipeline_constants.TWIN_PATCHES
+
+        assert stage.last_version_seen is None
+        stage.run_op(op)
+
+        assert stage.last_version_seen == -1
+
+    @pytest.mark.parametrize(
+        "feature_name",
+        [
+            pipeline_constants.C2D_MSG,
+            pipeline_constants.INPUT_MSG,
+            pipeline_constants.METHODS,
+            pipeline_constants.TWIN,
+        ],
+    )
+    @pytest.mark.it(
+        "Does not change `last_version_seen` if `op.feature_name` is not 'twin_patches'"
+    )
+    def test_doesnt_set_last_version_seen(self, mocker, stage, op, feature_name):
+        op.feature_name = feature_name
+        stage.last_version_seen = mocker.MagicMock()
+
+        old_value = stage.last_version_seen
+        stage.run_op(op)
+
+        assert stage.last_version_seen == old_value
+
+    @pytest.mark.parametrize(
+        "feature_name",
+        [
+            pipeline_constants.C2D_MSG,
+            pipeline_constants.INPUT_MSG,
+            pipeline_constants.METHODS,
+            pipeline_constants.TWIN,
+            pipeline_constants.TWIN_PATCHES,
+        ],
+    )
+    @pytest.mark.it(
+        "Sends the EnableFeatureOperation op to the next stage for all valid `op.feature_name` values"
+    )
```

Comment:
Probably not really practical to keep this always up to date, but not much to really do about it I suppose.